dss/events/handlers/index.py

```diff
@@ -45,43 +45,98 @@ def process_new_gs_indexable_object(event, logger) -> None:
+    blobstore = Config.get_cloud_specific_handles(replica)[0]
+
     if is_bundle_to_index(key):
-        logger.info(f"Received {replica} creation event for bundle which will be indexed: {key}")
-        blobstore = Config.get_cloud_specific_handles(replica)[0]
-        manifest = read_bundle_manifest(blobstore, bucket_name, key, logger)
-        bundle_id = get_bundle_id_from_key(key)
-        index_data = create_index_data(blobstore, bucket_name, bundle_id, manifest, logger)
-        alias_name = Config.get_es_alias_name(ESIndexType.docs, Replica[replica])
-        index_shape_identifier = get_index_shape_identifier(index_data, logger)
-        index_name = Config.get_es_index_name(ESIndexType.docs, Replica[replica], index_shape_identifier)
-        get_elasticsearch_index(index_name, alias_name, logger)
+        bundle_id = bundle_key_to_bundle_fqid(key)
+        index_name, alias_name, index_data = prepare_index(blobstore, replica, bucket_name, bundle_id, logger)
         add_data_to_elasticsearch(bundle_id, index_data, index_name, replica, logger)
-        subscriptions = find_matching_subscriptions(index_data, index_name, logger)
-        process_notifications(bundle_id, subscriptions, replica, logger)
-        logger.debug(f"Finished index processing of {replica} creation event for bundle: {key}")
+        notify_subscribers(bundle_id, index_data, index_name, replica, logger)
+    elif is_deletion(key):
+        delete_bundle(blobstore, replica, bucket_name, key, logger)
     else:
-        logger.debug(f"Not indexing {replica} creation event for key: {key}")
+        logger.debug(f"Not processing {replica} event for key: {key}")
+
+
+def delete_bundle(blobstore: BlobStore, replica: str, bucket_name: str, tombstone_object_name: str, logger):
+    bundle_uuid, version = DSS_OBJECT_NAME_REGEX.search(tombstone_object_name).groups()
+    if version:
+        # if a version is specified, delete just that version
+        bundle_fqid = format_bundle_fqid(bundle_uuid, version)
+        bundles_to_delete = [bundle_fqid]
+    else:
+        # if no version is specified, delete all bundle versions from the index
+        prefix = f"bundles/{bundle_uuid}."
+        bundles_to_delete = list(set([
+            bundle_key_to_bundle_fqid(k) for k in blobstore.list(bucket_name, prefix)
+            if DSS_BUNDLE_KEY_REGEX.match(k)
+        ]))
+
+    if not bundles_to_delete:
+        logger.info(f"No deletions processed for: {tombstone_object_name}")
+        return
+
+    tombstone_data = json.loads(blobstore.get(bucket_name, tombstone_object_name))
+
+    for bundle_fqid in bundles_to_delete:
+        index_name, _, _ = prepare_index(blobstore, replica, bucket_name, bundle_fqid, logger)
+        add_data_to_elasticsearch(bundle_fqid, tombstone_data, index_name, replica, logger)
+        logger.info(f"Deleted bundle {bundle_fqid}")
+
+
+def prepare_index(blobstore: BlobStore, replica: str, bucket_name: str, bundle_id: str, logger):
```

Comment:
This should probably be a method on BundleDocument as well. I am going to introduce one just like it in my next PR. A better name would be prepare_for_indexing.

Reply:
I actually stuck with your name (I didn't get it at first but once I saw what it actually does, the makes perfect sense):

https://github.com/HumanCellAtlas/data-store/pull/726/files#diff-5fee0aa8f4b4af997aab34ba08c78e22R153

Reply:
I wasn't totally satisfied with my naming, but glad you like it. I think your refactor is good, it brought some good mental structure here. I will leave this change to you in your other PR.